appcelerator/titanium_mobile
# fix(android): Improved merge of "tiapp.xml" android manifest settings

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26778
- https://jira.appcelerator.org/browse/TIMOB-26777
- https://jira.appcelerator.org/browse/TIMOB-27067
- https://jira.appcelerator.org/browse/TIMOB-27084

**Summary:**
- [TIMOB-26778](https://jira.appcelerator.org/browse/TIMOB-26778) Fixed bug where overriding an `<activity/>` in "tiapp.xml" caused most "configChanges" values to be lost.
  * Would cause activity window's UI to disappear if a system config change occurred dynamically.
- [TIMOB-26777](https://jira.appcelerator.org/browse/TIMOB-26777) Fixed bug where connecting/disconnecting a physical keyboard to/from device caused UI to disappear.
- [TIMOB-27067](https://jira.appcelerator.org/browse/TIMOB-27067) Fixed bug where UI sometimes disappears on Android 9.0 or higher when battery saver turns on/off.
- [TIMOB-27084](https://jira.appcelerator.org/browse/TIMOB-27084) Fixed issue where "tiapp.xml" was unable to override "AndroidManifest.xml" settings defined in AAR and "timodule.xml".
- Added missing `<activity/>` "configChanges" values to Titanium's template:
  * `keyboard`, `layoutDirection`, `mcc`, `mnc`, `navigation`, `touchscreen`, `uiMode`
- Added new EJS template variable `<%- allActivityConfigChanges %>`.
  * To be used by Titanium's main "AndroidManifest.xml" and "timodule.xml" files.
- Removed last remnants of deprecate `Ti.Map` from build script.
- Removed `TiVideoActivity` and `TiCameraActivity` injection code from build script.
  * There is no downside to always adding these activities to "AndroidManifest.xml". This reduces complexity.
- Removed launchMode "singleTask" from "TiMapActivity" and "TiVideoActivity".
  * Not applicable to child activities. Only applies to root activity.

---
**Test 1:**
1. Follow the test procedure attached to [TIMOB-26778](https://jira.appcelerator.org/browse/TIMOB-26778).
2. Verify that Titanium does not restart when rotating the splash screen window.
3. Verify that the generated "AndroidManifest.xml" file under the "build/android" directory has all of the following "configChanges" values.
```xml
<activity android:configChanges="fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode|density"/>
```

---
**Test 2:**
1. Follow the test procedure attached to [TIMOB-26777](https://jira.appcelerator.org/browse/TIMOB-26777).
2. Verify that connecting/disconnecting a bluetooth keyboard does **NOT** cause the UI to disappear.

---
**Test 3:**
1. Follow the test procedure attached to [TIMOB-27067](https://jira.appcelerator.org/browse/TIMOB-27067).
2. Verify that the window UI does not disappear when changing the battery "Charge connection" setting in the Android emulator.

---
**Test 4:**
1. Follow the test procedure attached to [TIMOB-27084](https://jira.appcelerator.org/browse/TIMOB-27084).
2. Verify that the "intent-filter" defined in your "tiapp.xml" appears in the "AndroidManifest.xml" file under the project's `./build/android` directory.
